File: gas/salesforce_notion_sync.js
/**
 * Salesforce → Notion 候補者DB 同期 GAS Web App
 *
 * 【アーキテクチャ】
 *   Salesforce Contact 登録・更新
 *     → Record-Triggered Flow が HTTP POST
 *     → この GAS Web App が受け取り
 *     → Notion 候補者DB に新規作成 or 更新
 *
 * 【セットアップ手順】
 *   1. Google Apps Script (script.google.com) で新規プロジェクト作成
 *   2. このコードを貼り付けて保存
 *   3. スクリプトプロパティを設定（下記参照）
 *   4. Web App としてデプロイ
 *   5. Salesforce Remote Site Settings に GAS URL を追加
 *   6. Salesforce Flow を設定（下記参照）
 *
 * 【スクリプトプロパティの設定】
 *   「プロジェクトの設定」→「スクリプトプロパティ」から以下を追加:
 *   - NOTION_TOKEN : Notion API トークン（secret_xxx...）
 *   - GAS_SECRET   : 任意の秘密キー（英数字で自由に設定、Salesforce Flow 側と合わせる）
 *
 * 【Web App デプロイ手順】
 *   「デプロイ」→「新しいデプロイ」
 *   - 種類: ウェブアプリ
 *   - 次のユーザーとして実行: 自分
 *   - アクセスできるユーザー: 全員
 *   → デプロイ後に表示される URL をコピーして Salesforce Flow に設定する
 *
 * 【Salesforce Flow 設定手順】
 *   1. フローを開く（設定 → フロー → 新規フロー）
 *   2. 「レコードトリガーフロー」を選択
 *   3. オブジェクト: 取引先責任者（Contact）
 *   4. トリガー: 「レコードが作成または更新された」
 *   5. 条件なし（全件対象）or 必要に応じてフィルタ設定
 *   6. アクションを追加 → 「HTTP コールアウト」
 *   7. 以下を設定:
 *      - URL    : GAS Web App の URL（デプロイ後に取得）
 *      - メソッド: POST
 *      - ヘッダー: Content-Type = application/json
 *      - 本文   : 下記 JSON テンプレートを使用
 *   8. Salesforce 設定 → リモートサイトの設定 に以下を追加:
 *      - リモートサイトのURL: https://script.google.com
 *
 * 【Flow の本文（JSON テンプレート）】
 *   以下を Flow の HTTP コールアウト本文に設定してください。
 *   {!変数名} の部分は Flow のマージ項目で置き換えます。
 *
 *   {
 *     "secret": "ここに GAS_SECRET の値を直接記入",
 *     "id": "{!$Record.Id}",
 *     "instanceUrl": "https://your-domain.salesforce.com",
 *     "lastName": "{!$Record.LastName}",
 *     "firstName": "{!$Record.FirstName}",
 *     "birthdate": "{!$Record.Birthdate}",
 *     "title": "{!$Record.Title}",
 *     "currentSalary": "{!$Record.現在年収のAPIカスタムフィールド名__c}",
 *     "desiredSalary": "{!$Record.希望年収のAPIカスタムフィールド名__c}",
 *     "jobChangeReason": "{!$Record.転職理由のAPIカスタムフィールド名__c}",
 *     "sideJob": "{!$Record.副業希望のAPIカスタムフィールド名__c}"
 *   }
 *
 *   ※ カスタムフィールドがない項目は行ごと削除してください
 *   ※ instanceUrl はご自身の SF 組織の URL に固定値で記入してください
 */

// ─── 定数 ────────────────────────────────────────────────────────────────────

// Notion 候補者DB の database ID（ダッシュなし32文字）
// collection://2057d017-b6a0-80fc-9e8d-000b3b6ab37e → 以下の値
const CANDIDATE_DB_ID = '2057d017b6a080fc9e8d000b3b6ab37e';
const NOTION_API_BASE = 'https://api.notion.com/v1';


// ─── エントリーポイント ───────────────────────────────────────────────────────

/**
 * Salesforce Flow からの POST を受け取る
 */
function doPost(e) {
  try {
    const data = JSON.parse(e.postData.contents);

    // 秘密キー認証
    const secret = PropertiesService.getScriptProperties().getProperty('GAS_SECRET');
    if (secret && data.secret !== secret) {
      return jsonResponse({ status: 'error', message: 'Unauthorized' });
    }

    const result = syncToNotion(data);
    console.log(`同期完了: ${JSON.stringify(result)}`);
    return jsonResponse({ status: 'success', ...result });

  } catch (err) {
    console.error('エラー:', err.message);
    return jsonResponse({ status: 'error', message: err.message });
  }
}

/**
 * 動作確認用（ブラウザで GAS URL にアクセスして確認）
 */
function doGet() {
  return jsonResponse({ status: 'ok', message: 'Salesforce→Notion GAS Web App is running' });
}


// ─── メイン同期処理 ───────────────────────────────────────────────────────────

function syncToNotion(data) {
  const token = PropertiesService.getScriptProperties().getProperty('NOTION_TOKEN');
  if (!token) throw new Error('スクリプトプロパティに NOTION_TOKEN が設定されていません');

  const sfUrl = `${data.instanceUrl}/${data.id}`;
  const existingPageId = findExistingPage(sfUrl, token);
  const properties = buildProperties(data, sfUrl);

  if (existingPageId) {
    updatePage(existingPageId, properties, token);
    return { action: 'updated', pageId: existingPageId, sfUrl };
  } else {
    const pageId = createPage(properties, token);
    return { action: 'created', pageId, sfUrl };
  }
}


// ─── Notion API ───────────────────────────────────────────────────────────────

function notionHeaders(token) {
  return {
    'Authorization': `Bearer ${token}`,
    'Notion-Version': '2022-06-28',
    'Content-Type': 'application/json',
  };
}

/**
 * SalesForce URL で候補者DB を検索し、既存ページの ID を返す（なければ null）
 */
function findExistingPage(sfUrl, token) {
  const res = UrlFetchApp.fetch(
    `${NOTION_API_BASE}/databases/${CANDIDATE_DB_ID}/query`,
    {
      method: 'post',
      headers: notionHeaders(token),
      payload: JSON.stringify({
        filter: { property: 'SalesForce', url: { equals: sfUrl } },
        page_size: 1,
      }),
      muteHttpExceptions: true,
    }
  );
  const json = JSON.parse(res.getContentText());
  const results = json.results || [];
  return results.length > 0 ? results[0].id : null;
}

/**
 * SF の Contact データから Notion プロパティ辞書を構築する
 */
function buildProperties(data, sfUrl) {
  const props = {};

  // 名前（姓 + 名）
  const lastName  = (data.lastName  || '').trim();
  const firstName = (data.firstName || '').trim();
  const fullName  = [lastName, firstName].filter(Boolean).join(' ');
  if (fullName) {
    props['名前'] = { title: [{ text: { content: fullName } }] };
  }

  // SalesForce URL（照合キー・必須）
  props['SalesForce'] = { url: sfUrl };

  // 生年月日
  if (data.birthdate) {
    props['生年月日'] = { date: { start: data.birthdate } };
  }

  // 現在の肩書き
  if (data.title) {
    props['ポジション'] = { rich_text: [{ text: { content: data.title } }] };
  }

  // 現年収（円）
  if (data.currentSalary != null && data.currentSalary !== '') {
    props['現年収'] = { number: Number(data.currentSalary) };
  }

  // 最低希望年収（円）
  if (data.desiredSalary != null && data.desiredSalary !== '') {
    props['最低希望年収'] = { number: Number(data.desiredSalary) };
  }

  // 転職検討理由
  if (data.jobChangeReason) {
    props['転職検討理由'] = { rich_text: [{ text: { content: String(data.jobChangeReason) } }] };
  }

  // 副業希望
  if (data.sideJob != null && data.sideJob !== '') {
    props['副業希望'] = { checkbox: data.sideJob === true || data.sideJob === 'true' };
  }

  return props;
}

/**
 * 候補者DB に新規ページを作成し、page_id を返す
 */
function createPage(properties, token) {
  const res = UrlFetchApp.fetch(`${NOTION_API_BASE}/pages`, {
    method: 'post',
    headers: notionHeaders(token),
    payload: JSON.stringify({
      parent: { database_id: CANDIDATE_DB_ID },
      properties,
    }),
    muteHttpExceptions: true,
  });
  const json = JSON.parse(res.getContentText());
  if (!json.id) throw new Error(`Notion ページ作成失敗: ${res.getContentText()}`);
  return json.id;
}

/**
 * 既存ページを更新する（名前は上書きしない）
 */
function updatePage(pageId, properties, token) {
  const { '名前': _omit, ...updateProps } = properties;
  if (Object.keys(updateProps).length === 0) return;

  const res = UrlFetchApp.fetch(`${NOTION_API_BASE}/pages/${pageId}`, {
    method: 'patch',
    headers: notionHeaders(token),
    payload: JSON.stringify({ properties: updateProps }),
    muteHttpExceptions: true,
  });
  if (res.getResponseCode() >= 400) {
    throw new Error(`Notion ページ更新失敗: ${res.getContentText()}`);
  }
}


// ─── ユーティリティ ───────────────────────────────────────────────────────────

function jsonResponse(data) {
  return ContentService
    .createTextOutput(JSON.stringify(data))
    .setMimeType(ContentService.MimeType.JSON);
}


// ─── ローカルテスト用（GAS エディタから実行して動作確認） ─────────────────────

/**
 * GAS エディタの「実行」ボタンで動作テストする場合はここを実行。
 * 実際の Salesforce データをハードコードして Notion への書き込みを確認できる。
 */
function testSync() {
  const testData = {
    secret: PropertiesService.getScriptProperties().getProperty('GAS_SECRET'),
    id: '0031234567890ABC',
    instanceUrl: 'https://your-domain.salesforce.com',  // ← 実際の URL に変更
    lastName: '山田',
    firstName: '太郎',
    birthdate: '1990-05-15',
    title: 'エンジニアリングマネージャー',
    currentSalary: 8000000,
    desiredSalary: 10000000,
    jobChangeReason: 'キャリアアップのため',
    sideJob: false,
  };

  const result = syncToNotion(testData);
  console.log('テスト結果:', JSON.stringify(result));
}

File: scripts/salesforce_notion_sync.py
"""
Salesforce 候補者 → Notion 候補者DB 同期スクリプト

【概要】
Salesforce の Contact レコード（候補者）を Notion 候補者DB に同期する。
・新規レコード  : Notion に新規ページとして作成する
・既存レコード  : --update フラグ付きで上書き更新する（SalesForce URL で照合）

【Salesforce フィールド設定】
スクリプト内の「カスタムフィールド設定」セクション、または .env の
SF_FIELD_* 変数をご自身の SF 組織のフィールド名に合わせて編集してください。

【セットアップ】
  pip install -r requirements.txt
  cp .env.example .env
  # NOTION_TOKEN, SF_USERNAME, SF_PASSWORD, SF_SECURITY_TOKEN を .env に記入

【実行オプション】
  --dry-run         Notion を更新せず確認のみ
  --update          既存の Notion ページも更新する（デフォルト: 新規のみ）
  --force           前回同期日時を無視して全件処理（デフォルト: 7日前以降）
  --debug           詳細ログ出力
  --since DATETIME  この日時以降の更新を対象にする（例: 2026-03-01T00:00:00）

【cron 例】毎時実行
  0 * * * * cd /path/to/pnl && python scripts/salesforce_notion_sync.py >> logs/sf_sync.log 2>&1
"""

import asyncio
import json
import os
import sys
from datetime import datetime, timedelta, timezone
from pathlib import Path
from zoneinfo import ZoneInfo

import httpx
from dotenv import load_dotenv
from simple_salesforce import Salesforce

# ─── 設定 ────────────────────────────────────────────────────────────────────

load_dotenv(Path(__file__).parent.parent / ".env")

NOTION_TOKEN    = os.environ["NOTION_TOKEN"]
# 候補者DB の Notion database ID
# collection://2057d017-b6a0-80fc-9e8d-000b3b6ab37e → ダッシュなし32文字
CANDIDATE_DB_ID = os.environ.get("CANDIDATE_DB_ID", "2057d017b6a080fc9e8d000b3b6ab37e")

SF_USERNAME       = os.environ["SF_USERNAME"]
SF_PASSWORD       = os.environ["SF_PASSWORD"]
SF_SECURITY_TOKEN = os.environ.get("SF_SECURITY_TOKEN", "")
SF_DOMAIN         = os.environ.get("SF_DOMAIN", "login")  # sandbox の場合は "test"

STATE_FILE = Path(__file__).parent / "sf_sync_state.json"
JST = ZoneInfo("Asia/Tokyo")

NOTION_API = "https://api.notion.com/v1"
NOTION_HEADERS = {
    "Authorization": f"Bearer {NOTION_TOKEN}",
    "Notion-Version": "2022-06-28",
    "Content-Type": "application/json",
}

DRY_RUN = "--dry-run" in sys.argv
UPDATE   = "--update"  in sys.argv
FORCE    = "--force"   in sys.argv
DEBUG    = "--debug"   in sys.argv

# --since DATETIME 引数のパース
_SINCE_ARG: str | None = None
for _i, _arg in enumerate(sys.argv):
    if _arg == "--since" and _i + 1 < len(sys.argv):
        _SINCE_ARG = sys.argv[_i + 1]
        break


# ─── Salesforce カスタムフィールド設定 ────────────────────────────────────────
#
# ご自身の Salesforce 組織のカスタムフィールド API 参照名に合わせて変更してください。
# .env の SF_FIELD_* 変数で上書きも可能です。
#
# フィールドが SF 組織に存在しない場合、自動的にスキップされます。
#
# 対応する Notion 候補者DB フィールド:
#   現年収 (number)         ← SF_FIELD_CURRENT_SALARY
#   最低希望年収 (number)   ← SF_FIELD_DESIRED_SALARY
#   転職検討理由 (text)     ← SF_FIELD_JOB_CHANGE_REASON
#   副業希望 (checkbox)     ← SF_FIELD_SIDE_JOB

SF_FIELD_CURRENT_SALARY    = os.environ.get("SF_FIELD_CURRENT_SALARY",    "Current_Salary__c")
SF_FIELD_DESIRED_SALARY    = os.environ.get("SF_FIELD_DESIRED_SALARY",    "Desired_Salary__c")
SF_FIELD_JOB_CHANGE_REASON = os.environ.get("SF_FIELD_JOB_CHANGE_REASON", "Job_Change_Reason__c")
SF_FIELD_SIDE_JOB          = os.environ.get("SF_FIELD_SIDE_JOB",          "Side_Job_Desired__c")


# ─── 状態管理 ─────────────────────────────────────────────────────────────────

def load_state() -> dict:
    if STATE_FILE.exists():
        return json.loads(STATE_FILE.read_text())
    return {}


def save_state(last_sync: str, synced_count: int) -> None:
    STATE_FILE.write_text(
        json.dumps({"last_sync": last_sync, "synced_count": synced_count},
                   ensure_ascii=False, indent=2)
    )


def get_since_datetime() -> str:
    """同期対象の起点日時（ISO8601）を返す"""
    if _SINCE_ARG:
        return _SINCE_ARG if "T" in _SINCE_ARG else f"{_SINCE_ARG}T00:00:00Z"

    if not FORCE:
        state = load_state()
        if state.get("last_sync"):
            return state["last_sync"]

    # 初回 or --force: 7日前から
    dt = datetime.now(timezone.utc) - timedelta(days=7)
    return dt.strftime("%Y-%m-%dT%H:%M:%SZ")


# ─── Salesforce: 候補者取得 ────────────────────────────────────────────────────

def connect_salesforce() -> Salesforce:
    return Salesforce(
        username=SF_USERNAME,
        password=SF_PASSWORD,
        security_token=SF_SECURITY_TOKEN,
        domain=SF_DOMAIN,
    )


def build_sf_record_url(instance_url: str, record_id: str) -> str:
    return f"{instance_url}/{record_id}"


def fetch_candidates(sf: Salesforce, since: str) -> tuple[list[dict], list[str]]:
    """
    Salesforce から候補者（Contact）を取得する。

    - LastModifiedDate が since 以降のレコードを対象
    - 存在しないカスタムフィールドは自動スキップ
    - 戻り値: (records, available_custom_fields)
    """
    base_fields = ["Id", "FirstName", "LastName", "Birthdate", "Title"]
    custom_field_candidates = [
        SF_FIELD_CURRENT_SALARY,
        SF_FIELD_DESIRED_SALARY,
        SF_FIELD_JOB_CHANGE_REASON,
        SF_FIELD_SIDE_JOB,
    ]

    # 組織に存在するフィールドのみ選択
    all_field_names = {f["name"] for f in sf.Contact.describe()["fields"]}
    available_custom: list[str] = []
    select_fields = base_fields.copy()

    for cf in custom_field_candidates:
        if cf in all_field_names:
            select_fields.append(cf)
            available_custom.append(cf)
        elif DEBUG:
            print(f"  [DEBUG] SF フィールド未検出（スキップ）: {cf!r}")

    fields_str = ", ".join(select_fields)
    soql = (
        f"SELECT {fields_str} "
        f"FROM Contact "
        f"WHERE LastModifiedDate >= {since} "
        f"ORDER BY LastModifiedDate ASC "
        f"LIMIT 500"
    )

    if DEBUG:
        print(f"  [DEBUG] SOQL: {soql}")

    result = sf.query_all(soql)
    return result.get("records", []), available_custom


# ─── Notion: 候補者DB 操作 ─────────────────────────────────────────────────────

async def find_existing_page(sf_url: str, client: httpx.AsyncClient) -> str | None:
    """SalesForce URL で候補者DB を検索。存在すれば page_id を返す"""
    payload = {
        "filter": {
            "property": "SalesForce",
            "url": {"equals": sf_url},
        },
        "page_size": 1,
    }
    res = await client.post(
        f"{NOTION_API}/databases/{CANDIDATE_DB_ID}/query",
        headers=NOTION_HEADERS,
        json=payload,
    )
    res.raise_for_status()
    results = res.json().get("results", [])
    return results[0]["id"] if results else None


def build_notion_properties(
    record: dict,
    sf_url: str,
    available_custom: list[str],
) -> dict:
    """
    SF Contact レコードから Notion プロパティ辞書を構築する。
    値が None のフィールドはプロパティに含めない（Notion の既存値を保持）。
    """
    last_name  = (record.get("LastName")  or "").strip()
    first_name = (record.get("FirstName") or "").strip()
    full_name  = f"{last_name} {first_name}".strip() or None

    props: dict = {}

    if full_name:
        props["名前"] = {"title": [{"text": {"content": full_name}}]}

    props["SalesForce"] = {"url": sf_url}

    if record.get("Birthdate"):
        props["生年月日"] = {"date": {"start": record["Birthdate"]}}

    if record.get("Title"):
        props["ポジション"] = {"rich_text": [{"text": {"content": record["Title"]}}]}

    if SF_FIELD_CURRENT_SALARY in available_custom:
        val = record.get(SF_FIELD_CURRENT_SALARY)
        if val is not None:
            props["現年収"] = {"number": int(val)}

    if SF_FIELD_DESIRED_SALARY in available_custom:
        val = record.get(SF_FIELD_DESIRED_SALARY)
        if val is not None:
            props["最低希望年収"] = {"number": int(val)}

    if SF_FIELD_JOB_CHANGE_REASON in available_custom:
        val = record.get(SF_FIELD_JOB_CHANGE_REASON)
        if val:
            props["転職検討理由"] = {"rich_text": [{"text": {"content": str(val)}}]}

    if SF_FIELD_SIDE_JOB in available_custom:
        val = record.get(SF_FIELD_SIDE_JOB)
        if val is not None:
            props["副業希望"] = {"checkbox": bool(val)}

    return props


async def create_notion_page(properties: dict, client: httpx.AsyncClient) -> str:
    """候補者DB に新規ページを作成して page_id を返す"""
    payload = {
        "parent": {"database_id": CANDIDATE_DB_ID},
        "properties": properties,
    }
    res = await client.post(
        f"{NOTION_API}/pages",
        headers=NOTION_HEADERS,
        json=payload,
    )
    res.raise_for_status()
    return res.json()["id"]


async def update_notion_page(
    page_id: str, properties: dict, client: httpx.AsyncClient
) -> None:
    """Notion ページのプロパティを更新する（名前は更新しない）"""
    update_props = {k: v for k, v in properties.items() if k != "名前"}
    if not update_props:
        return
    res = await client.patch(
        f"{NOTION_API}/pages/{page_id}",
        headers=NOTION_HEADERS,
        json={"properties": update_props},
    )
    res.raise_for_status()


# ─── メイン処理 ───────────────────────────────────────────────────────────────

async def main() -> None:
    now = datetime.now(tz=JST)
    ts  = now.strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{ts}] Salesforce → Notion 候補者DB 同期開始")

    if DRY_RUN:
        print("⚠️  --dry-run モード: Notion の更新は行いません")
    if UPDATE:
        print("ℹ️  --update モード: 既存レコードも更新します")

    since = get_since_datetime()
    print(f"📅 同期対象: {since} 以降に更新されたレコード")

    # ① Salesforce から候補者を取得
    print("\n📥 Salesforce から候補者を取得中...")
    try:
        sf = connect_salesforce()
        records, available_custom = fetch_candidates(sf, since)
        instance_url = sf.base_url.split("/services")[0]
    except Exception as e:
        print(f"❌ Salesforce 接続/取得失敗: {e}")
        sys.exit(1)

    if not records:
        print("ℹ️  新規・更新された候補者が見つかりませんでした。")
        if not DRY_RUN:
            save_state(now.strftime("%Y-%m-%dT%H:%M:%SZ"), 0)
        return

    print(f"   {len(records)} 件取得")
    if available_custom and DEBUG:
        print(f"   [DEBUG] 利用可能なカスタムフィールド: {available_custom}")

    # ② Notion と照合して作成/更新
    print(f"\n🔄 Notion 候補者DB に同期中...")

    created_count = 0
    updated_count = 0
    skipped_count = 0

    async with httpx.AsyncClient(timeout=30) as client:
        for record in records:
            record_id = record["Id"]
            sf_url    = build_sf_record_url(instance_url, record_id)
            last_name  = (record.get("LastName")  or "").strip()
            first_name = (record.get("FirstName") or "").strip()
            name = f"{last_name} {first_name}".strip() or record_id

            if DEBUG:
                print(f"  [DEBUG] 処理中: {name!r} ({sf_url})")

            try:
                existing_page_id = await find_existing_page(sf_url, client)
            except Exception as e:
                print(f"  ⚠️  Notion 検索失敗 ({name}): {e}")
                skipped_count += 1
                continue

            properties = build_notion_properties(record, sf_url, available_custom)

            if existing_page_id:
                if UPDATE:
                    if DRY_RUN:
                        print(f"  [dry-run] 更新: {name!r}")
                    else:
                        try:
                            await update_notion_page(existing_page_id, properties, client)
                            print(f"  ✅ 更新: {name!r}")
                            updated_count += 1
                        except Exception as e:
                            print(f"  ⚠️  更新失敗 ({name}): {e}")
                            skipped_count += 1
                else:
                    if DEBUG:
                        print(f"  [DEBUG] スキップ（既存）: {name!r}")
                    skipped_count += 1
            else:
                if DRY_RUN:
                    print(f"  [dry-run] 新規作成: {name!r}")
                else:
                    try:
                        await create_notion_page(properties, client)
                        print(f"  ✅ 新規作成: {name!r}")
                        created_count += 1
                    except Exception as e:
                        print(f"  ⚠️  作成失敗 ({name}): {e}")
                        skipped_count += 1

    # ③ 状態保存
    if not DRY_RUN:
        save_state(now.strftime("%Y-%m-%dT%H:%M:%SZ"), created_count + updated_count)

    ts_end = datetime.now(tz=JST).strftime("%Y-%m-%d %H:%M:%S")
    print(
        f"\n[{ts_end}] 完了: "
        f"新規作成={created_count}, 更新={updated_count}, スキップ={skipped_count}"
    )


if __name__ == "__main__":
    asyncio.run(main())
